virtio-devices/src/iommu.rs

```diff
@@ -563,15 +572,25 @@ impl Request {
                         }
                     }
 
-                    // Remove mapping associated with the domain
-                    mapping
-                        .domains
-                        .write()
-                        .unwrap()
-                        .get_mut(&domain_id)
-                        .unwrap()
-                        .mappings
-                        .remove(&virt_start);
+                    // Remove all mappings associated with the domain
+                    let mut size = req.virt_end - virt_start + 1;
+                    let mut virt_start = req.virt_start;
+                    while size > 0 {
+                        if let Some(addr_map) = mapping
+                            .domains
+                            .write()
+                            .unwrap()
+                            .get_mut(&domain_id)
+                            .unwrap()
+                            .mappings
+                            .remove(&virt_start)
+                        {
+                            size -= addr_map.size;
+                            virt_start += addr_map.size;
+                        } else {
+                            break;
+                        }
+                    }
```

Comment:
I see why you are doing this. The iommu specification allows a UNMAP request to unmap a region of memory that contains multiple mappings. However, this logic is still not quite correct. The iommu spec does not require that the entire region to be unmapped has been previously mapped. The only restriction on UNMAP is that it cannot be used to partially unmap a previous mapping. For example, according to the spec it is possible to remove all mappings simply by creating a UNMAP request on (0, 1 << phys_bits). 

It would be more correct to iterate through all mappings, and remove all mappings that fall entirely within the region `(virt_start, virt_end)`.

Reply:
Sure additional PR here: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/6297